### ISSUE TITLE
Add Neural Net Learner in learning.ipynb

### ISSUE THREAD

**Issue by CharuChhimpa** [aimacode/aima-python#911]
@MrDupin and @norvig, Can I add Neural Net Learner in learning.ipynb?

**Comment by MrDupin** [comment]
Sorry, we already have a section on Neural Nets, on [neural_nets.ipynb](https://github.com/aimacode/aima-python/blob/master/neural_nets.ipynb). You can work more on that if you want, expanding and it and adding visualizations/examples.

**Comment by uchihabloodline** [comment]
> Sorry, we already have a section on Neural Nets, on [neural_nets.ipynb](https://github.com/aimacode/aima-python/blob/master/neural_nets.ipynb). You can work more on that if you want, expanding and it and adding visualizations/examples.

Can I add more implementations and algorithms regarding neural Nets and related algorithms in this section. I have been studying Deep learning since a while and now want to contribute regarding the same as I will be applying for the GSOC this year.

**Comment by MrDupin** [comment]
@uchihabloodline: Depends on the algorithm. If it is a simple enough implementation, you are more than welcome to try it!

Remember that we want to keep the project as devoid of external libraries as possible.

**Comment by uchihabloodline** [comment]
> @uchihabloodline: Depends on the algorithm. If it is a simple enough implementation, you are more than welcome to try it!
> 
> Remember that we want to keep the project as devoid of external libraries as possible.

Yes, mostly I am gonna implement most of the algorithms using the usual libraries of python(say Numpy) but some of them might contain some high level API like keras, tensorflow or pytorch but I,ll try to make sure this does'nt happen. Will it be fine for proposing abstract of project in GSOC-19??

**Comment by MrDupin** [comment]
@uchihabloodline: I don't think keras/tensorflow/etc. are libraries we want to be using in this project, since their installation is not always straightforward and some people may not want these libraries. This project is more for educational purposes, providing background on some algorithms.

Having said that, Dr. Norvig did mention once that he wants to start integrating tensorflow in this project, but there hasn't been an update on that.

**Comment by uchihabloodline** [comment]
> @uchihabloodline: I don't think keras/tensorflow/etc. are libraries we want to be using in this project, since their installation is not always straightforward and some people may not want these libraries. This project is more for educational purposes, providing background on some algorithms.
> 
> Having said that, Dr. Norvig did mention once that he wants to start integrating tensorflow in this project, but there hasn't been an update on that.

Ok, I,ll make sure that every algorithm I propose will be implemented from scratch using the usual libraries of python only. From Variational auto encoders to Boltzmann machine.

**Comment by abbas5253** [comment]
@MrDupin and @norvig can add optimization algorithm for gradient desent in Backproplearner() such as Adam Optimizer option in the neural net in learning.py which takes extra argument in NeuralNetlearner() as Optimization_for_gradientDescent = True/False and can function accordingly...

**Comment by MrDupin** [comment]
Hello @abbas5253, you can definitely add that, great idea! Just remember to keep the code as simple as possible with no use of external libraries.

**Comment by abbas5253** [comment]
ok I'll make sure that no external library will be used

**Comment by dmeoli** [comment]
A neural-network learner is covered in `neural_nets.ipynb` (NEURAL NETWORK ALGORITHM + BACKPROPAGATION sections with a worked `NeuralNetLearner(iris)` example), and there's further deep-learning coverage in `notebooks/chapter19/` (Learners / Backpropagation / Loss-Functions notebooks). `learning.ipynb` itself covers the Perceptron classifier. Closing as covered.
